### PR TITLE
[AArch64] Add flag setting instructions to scheduling model.

### PR DESCRIPTION
**Pull request by Asher8118**
Some flag setting instructions (such as ANDS, ADDS, CCMN) were missing from the V2 scheduling model. This patch adds them in.

### CODE REVIEW

**Review by sjoerdmeijer — COMMENTED** [review]
Bit of nit, but I don't see test changes for ADC, SBC, and BIC?

**Comment by rin-arm** [comment]
> Bit of nit, but I don't see test changes for ADC, SBC, and BIC?

ADC and SBC had existing tests that were not influenced by this change, as both were already defined. It looks like BIC was also defined already so I removed it from the patch.

**Review by sjoerdmeijer — APPROVED** [review]
Excellent, in that case it LGTM.

**Review by davemgreen — APPROVED** [review]

File: llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td [hunk 1]
```diff
@@ -1108,12 +1108,18 @@ def : InstRW<[V2Write_1cyc_1B_1R], (instrs BL, BLR)>;
 // ALU, basic
 // ALU, basic, flagset
 def : SchedAlias<WriteI, V2Write_1cyc_1I>;
-def : InstRW<[V2Write_1cyc_1F], (instregex "^(ADC|SBC)S[WX]r$")>;
+def : InstRW<[V2Write_1cyc_1F], (instregex "^(ADD|SUB)S[WX]r[ir]$",
```
Comment (rj-jesus) [inline llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1111]: I believe `r` is wrong in the last group, should be `s`, but that's already captured a few lines down.
```suggestion
def : InstRW<[V2Write_1cyc_1F], (instregex "^(ADD|SUB)S[WX]ri$",
```
Also, it is not clear to me if for ADD{S}/SUB{S} (immediate), this should only apply for "LSL shift, shift <= 4". Have you tested with `LSL #12`?

File: llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td [hunk 2]
```diff
@@ -1108,12 +1108,18 @@ def : InstRW<[V2Write_1cyc_1B_1R], (instrs BL, BLR)>;
...
+                        "^(ADC|SBC)S[WX]r$",
+                        "^ANDS[WX]ri$")>;
```
Comment (rj-jesus) [inline llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1113]: We don't seem to have tests for the ANDS---could you please add a couple?

File: llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td [hunk 3]
```diff
@@ -1108,12 +1108,18 @@ def : InstRW<[V2Write_1cyc_1B_1R], (instrs BL, BLR)>;
...
 def : InstRW<[V2Write_0or1cyc_1I], (instregex "^MOVZ[WX]i$")>;
 
 // ALU, extend and shift
 def : SchedAlias<WriteIEReg, V2Write_2cyc_1M>;
 
+// Conditional compare
+def : InstRW<[V2Write_1cyc_1F],
+            (instregex "^CCMP(W|X)(i|r)", "^CCMN(W|X)(i|r)")>;
```
Comment (rj-jesus) [inline llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1121]: Could this be collapsed as below?
```suggestion
def : InstRW<[V2Write_1cyc_1F], (instregex "^CCM[NP][WX][ir]")>;
```

**Review by rj-jesus — COMMENTED** [review]
Thanks for this, just have a couple of nits. Also, did you test this for ADD{S}/SUB{S} (immediate) with `LSL #12`? Based on the SOG it's unclear to me if this case should have a latency and throughput of 2 and use pipeline M. Otherwise, LGTM.

File: llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1111]: ADDSWrr is a psuedo that is used during codegen and turned into a ADDSWrs with a zero shift. I believe it should be treated like a (cheap) ADDSWrs for scheduling.

And for ri I believe they are constants either way, so does not apply to the same "free vs expensive shifts" as in the ADDrs variants. i.e they are always "ALU, basic, flagset".

Comment (rj-jesus) [reply llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1111]: I see, perfect, thanks for clarifying.

File: llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td (hunk 2 above)
Comment (rj-jesus) [reply llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:1113]: Thanks

**Review by rj-jesus — APPROVED** [review]

**Pull request merged**